Add Wrangler v4 RC changelog entry (#20355)
cloudflare/cloudflare-docs: src/content/changelog/workers/2025-02-27-wrangler-v4-rc.mdx

---

> title: Introducing Wrangler v4's release candidate

Author: ```suggestion
title: Introducing Wrangler's v4 release candidate
```

description: You can now try out the v4 RC for Wrangler by installing `wrangler@4.0.0-rc.0`
products:
  - workers
date: 2025-02-27T11:00:00Z
---

import { PackageManagers } from "~/components";

TLDR: Try out the Wrangler v4 RC by installing via your favourite package manager:

<PackageManagers pkg="wrangler@v4-rc" />

> Wrangler v3 was released in May 2023, and since then has seen amazing growth and usage—from just \~250,000 downloads a week back then to 1M+ now. It's time for the next chapter in Wrangler's journey: v4. This release will bring new features and unlock new possibilities for the developer platform, but for the vast majority of users you _won't need to make any changes to upgrade_. It'll be as simple as upgrading Wrangler.

Reviewer: ```suggestion
Wrangler v3 was released in May 2023, and since then has seen amazing growth and usage—from just \~250,000 downloads a week back then to 1M+ now. It's time for the next chapter in Wrangler's journey: v4. This release will bring new features and unlock new possibilities for the developer platform. And for the vast majority of users, _no changes will be necessary to upgrade_ -- it'll be as simple as upgrading Wrangler.
```

> We're working on a more detailed migration guide, but here's a brief list of the changes to expect:

Reviewer: ```suggestion
We're working on a detailed migration guide that will accompany the v4 release. In the meantime, here's a brief list of the changes to expect:
```

#### Changes

- **Upgraded `esbuild` version**

>   This has been a highly-requested feature for a long time, and will allow you to use more modern JavaScript features in your Workers, including import attributes and `using` statements

Reviewer: ```suggestion
  A long-time highly-requested feature, this will now allow the use of more modern JavaScript features in Workers, including import attributes and `using` statements
```

- **`--local` by default for all commands**
  All commands that access resources (e.g. `wrangler kv`, `wrangler r2`, `wrangler d1`) will now access local datastores by default

- **Updated Node.js support policy**
  Wrangler will now support the active, maintenance, and current versions of Node.js (currently v18+)

- **Deprecated features removed**

>   All deprecated features have been removed from Wrangler, including

Reviewer: ```suggestion
  All deprecated features will be removed from Wrangler, including
```

this it to keep the tensing the same with items above

You can run the following command to install it and be one of the first to try it out:

<PackageManagers pkg="wrangler@v4-rc" />

Unlike previous major versions of Wrangler, which were [foundational rewrites](https://blog.cloudflare.com/wrangler-v2-beta/) and [rearchitectures](https://blog.cloudflare.com/wrangler3/) — Version 4 of Wrangler includes a much smaller set of changes. If you use Wrangler today, your workflow is very unlikely to change. Before we release Wrangler v4 and advance past the release candidate stage, we'll share a detailed migration guide in the Workers developer docs. But for the vast majority of cases, you won't need to do anything to migrate — things will just work as they do today. We are sharing this release candidate in advance of the official release of v4, so that you can try it out early and share feedback.

#### New JavaScript language features that you can now use with Wrangler v4

Version 4 of Wrangler updates the version of [esbuild](https://esbuild.github.io/) that Wrangler uses internally, allowing you to use modern JavaScript language features, including:

##### The `using` keyword from Explicit Resource Management

> The [`using` keyword from the Explicit Resource Management standard](/workers/runtime-apis/rpc/lifecycle/#explicit-resource-management), making it easier to work with the [JavaScript-native RPC system built into Workers](/workers/runtime-apis/rpc/). This means that when you obtain a stub, you can ensure that it is automatically disposed when you exit scope it was created in:

Reviewer: ```suggestion
The [`using` keyword from the Explicit Resource Management standard](/workers/runtime-apis/rpc/lifecycle/#explicit-resource-management) makes it easier to work with the [JavaScript-native RPC system built into Workers](/workers/runtime-apis/rpc/). This means that when you obtain a stub, you can ensure that it is automatically disposed when you exit scope it was created in:
```

```js
import data from "./data.json" with { type: "json" };
```

#### Other changes

##### `--local` is now the default for all CLI commands

All commands that access resources (for example, `wrangler kv`, `wrangler r2`, `wrangler d1`) now access local datastores by default, ensuring consistent behavior.

##### Clearer policy for the minimum required version of Node.js required to run Wrangler

Moving forward, the [active, maintenance, and current versions of Node.js](https://nodejs.org/en/about/previous-releases) will be officially supported by Wrangler. This means the minimum officially supported version of Node.js you must have installed for Wrangler v4 will be Node.js v18 or later. This policy mirrors how many other packages and CLIs support older versions of Node.js, and ensures that as long as you are using a version of Node.js that the Node.js project itself supports, this will be supported by Wrangler as well.

##### Features previously deprecated in Wrangler v3 are now removed in Wrangler v4

> All previously deprecated features in [Wrangler v2](https://developers.cloudflare.com/workers/wrangler/deprecations/#wrangler-v2) and in [Wrangler v3](https://developers.cloudflare.com/workers/wrangler/deprecations/#wrangler-v3) have now been removed. Additionally, the following features that were deprecated during the Wrangler v3 release lifecycle have been removed:

Reviewer: ```suggestion
All previously deprecated features in [Wrangler v2](https://developers.cloudflare.com/workers/wrangler/deprecations/#wrangler-v2) and in [Wrangler v3](https://developers.cloudflare.com/workers/wrangler/deprecations/#wrangler-v3) have now been removed. Additionally, the following features that were deprecated during the Wrangler v3 release have been removed:
```

removed "lifecycle" because don't know if it's necessary wording to include

```js
import data from "./data.json" with { type: "json" };
```

#### Other changes

##### `--local` is now the default for all CLI commands

All commands that access resources (for example, `wrangler kv`, `wrangler r2`, `wrangler d1`) now access local datastores by default, ensuring consistent behavior, intended to reduce confusion

##### Clearer policy for the minimum required version of Node.js required to run Wrangler

Moving forward, the [active, maintenance, and current versions of Node.js](https://nodejs.org/en/about/previous-releases) will be officially supported. This means the minimum officially supported version of Node.js for Wrangler v4 will e Node.js v18 or later. This policy mirrors how many other packages and CLIs support older versions of Node.js, and ensures that as long as you are using a version of Node.js that the Node.js project itself supports, this will be supported by Wrangler as well.

##### Features previously deprecated in Wrangler v3 are now removed in Wrangler v4

The following features, previously deprecated over one year ago in Wrangler v3, now can no longer be used in Wrangler v4:

(think these need a — instead do x...)

>   - Arguments:

Reviewer 2: @lrapoport-cf @penalosa for the things below, I think if you work on Wrangler it's easy to look at this list and know "oh yeah these are all old things that there are other ways of doing" — but right now this comes across as a long list without guidance on what people can do instead.

Especially since we already have a lot of the deprecations and guidance here:

https://developers.cloudflare.com/workers/wrangler/deprecations/#wrangler-v3

...should we just update that page with the RC v4 stuff, and in this post, link directly to it?

I think it could help us tighten up the post, and also help send the reader to the page that will give them the relevant details they need

Author: I included the list for completeness, but we could actually remove a fair few—a lot of these were deprecated in v2 and have never worked in v2+

Author: Sounds good, although maybe we shouldn't update the deprecation page at this point? Once v4 is released properly we'll have a full migration guide we can point people to, but that's not fully ready yet.

Author: I don't feel strongly about it though—happy to update that now if you think that makes sense.

Author: I've made this more user-focused in https://github.com/cloudflare/cloudflare-docs/pull/20355/commits/419f9c6a25f8e05f5041e4495d353cbccaec256a—how does that look?

Reviewer: i left a couple really small comments, but overall looks good to me! agree with @irvinebroque we should update the deprecations page, for completeness
